src/main/scala/analysis/Analysis.scala

```diff
@@ -147,73 +147,91 @@ object Fresh {
   }
 }
 
-trait MemoryRegion
-
-/** Represents a stack region. The region is defined by a region Identifier identifying the assignment location. There
-  * can exist two regions with the same size (offset) but have a different base pointer. As such the base pointer is
-  * tracked but not printed in the toString method.
-  * @param start
-  *   0x1234 in case of mem[R1 + 0x1234] <- ...
-  * @param regionType
-  *   The type of the region. This is used to distinguish between stack, heap, data and code regions.
-  * @param extent
-  *   the start and end of the region
-  */
-case class StackRegion(regionIdentifier: String, start: Expr, var extent: Option[RangeKey]) extends MemoryRegion:
-  override def toString: String = s"Stack(${regionIdentifier}, ${start})"
-  override def hashCode(): Int = regionIdentifier.hashCode()
+trait MemoryRegion {
+  val regionIdentifier: String
+  val start: BitVecLiteral
+  var extent: Option[RangeKey] = None
+}
+
+class StackRegion(override val regionIdentifier: String, override val start: BitVecLiteral) extends MemoryRegion {
+  override def toString: String = s"Stack($regionIdentifier, $start)"
+  override def hashCode(): Int = regionIdentifier.hashCode() * start.hashCode()
   override def equals(obj: Any): Boolean = obj match {
-    case StackRegion(ri, st, _) => st == start
-    case _                      => false
+    case s: StackRegion => s.start == start && s.regionIdentifier == regionIdentifier
+    case _ => false
   }
+}
 
-/** Represents a Heap region. The region is defined by its identifier which is defined by the allocation site.
-  * @param regionIdentifier
-  *   region id identifying the call-site
-  * @param start
-  *   the start address
-  * @param extent
-  *   the start and end of the region
-  */
-case class HeapRegion(regionIdentifier: String, start: Expr, var extent: Option[RangeKey]) extends MemoryRegion:
-  override def toString: String = s"Heap(${regionIdentifier}, ${start})"
-  override def hashCode(): Int = regionIdentifier.hashCode()
+class HeapRegion(override val regionIdentifier: String, override val start: BitVecLiteral) extends MemoryRegion {
+  override def toString: String = s"Heap($regionIdentifier, $start)"
+  override def hashCode(): Int = regionIdentifier.hashCode() * start.hashCode()
   override def equals(obj: Any): Boolean = obj match {
-    case r: HeapRegion => regionIdentifier.equals(r.regionIdentifier)
-    case _             => false
+    case h: HeapRegion => h.start == start && h.regionIdentifier == regionIdentifier
```

Comment:
I think the analysis makes start the size of the allocation so we don't want to use it in the region equality (different sizes should imply different identifiers)? But we might as well make these case classes?

Reply:
They were case classes before but I changed them to be classes because it was pointless to make them case classes when the hashCode and equals methods were being overwritten, and they had the mutable extent parameter. I'm just going to remove HeapRegion.start for now since it's misleading and incorrect (nothing meaningful is even done with the HeapRegions at present anyway) but fixing #72 will require a broader overhaul anyway.